File: sdk/go/examples/04-templates/sykli.go
//go:build ignore

// Example 04: Templates
//
// This example demonstrates:
// - Template() for reusable configurations
// - From() to inherit template settings
// - Override template settings per-task
//
// Templates eliminate repetition. Define container, mounts,
// and environment once, reuse across many tasks.
//
// Run with: sykli run

package main

import sykli "sykli.dev/go"

func main() {
	s := sykli.New()

	// === RESOURCES ===
	src := s.Dir(".")
	goModCache := s.Cache("go-mod")
	goBuildCache := s.Cache("go-build")

	// === TEMPLATES ===

	// Go template - common config for all Go tasks
	golang := s.Template("golang").
		Container("golang:1.21").
		Mount(src, "/src").
		MountCache(goModCache, "/go/pkg/mod").
		MountCache(goBuildCache, "/root/.cache/go-build").
		Workdir("/src").
		Env("GOFLAGS", "-mod=readonly")

	// Node template - for JS tooling
	node := s.Template("node").
		Container("node:20-slim").
		Mount(src, "/src").
		MountCache(s.Cache("npm"), "/root/.npm").
		Workdir("/src")

	// === TASKS ===

	// All Go tasks inherit from the golang template
	// Only task-specific config needed now!
	s.Task("lint").From(golang).Run("go vet ./...")
	s.Task("test").From(golang).Run("go test ./...")

	// Override template settings when needed
	s.Task("build").
		From(golang).
		Env("CGO_ENABLED", "0").  // Overrides/adds to template env
		Run("go build -o ./app").
		Output("binary", "./app").
		After("lint", "test")

	// Different template for JS tasks
	s.Task("docs").
		From(node).
		Run("npm run build:docs")

	s.Emit()
}

// Without templates (repetitive):
//   s.Task("lint").Container("golang:1.21").Mount(src, "/src")...
//   s.Task("test").Container("golang:1.21").Mount(src, "/src")...
//   s.Task("build").Container("golang:1.21").Mount(src, "/src")...
//
// With templates (DRY):
//   golang := s.Template("golang").Container("golang:1.21").Mount(src, "/src")...
//   s.Task("lint").From(golang).Run("go vet")
//   s.Task("test").From(golang).Run("go test")
//   s.Task("build").From(golang).Run("go build")

File: sdk/elixir/examples/04-templates/sykli.exs
# Example 04: Templates
#
# This example demonstrates:
# - `template` for reusable configurations
# - `from` to inherit template settings
# - Override template settings per-task
#
# Templates eliminate repetition. Define container, mounts,
# and environment once, reuse across many tasks.
#
# Run with: elixir sykli.exs --emit | sykli run -

Mix.install([{:sykli, "~> 0.2"}])

use Sykli

pipeline do
  # === RESOURCES ===
  src = dir(".")
  deps_cache = cache("mix-deps")
  build_cache = cache("mix-build")
  npm_cache = cache("npm-cache")

  # === TEMPLATES ===

  # Elixir template - common config for all Elixir tasks
  elixir = template("elixir")
    |> template_container("elixir:1.16")
    |> template_mount("src:.", "/app")
    |> template_mount_cache("mix-deps", "/app/deps")
    |> template_mount_cache("mix-build", "/app/_build")
    |> template_workdir("/app")

  # Node template - for JS tooling
  node = template("node")
    |> template_container("node:20-slim")
    |> template_mount("src:.", "/app")
    |> template_mount_cache("npm-cache", "/root/.npm")
    |> template_workdir("/app")

  # === TASKS ===

  # All Elixir tasks inherit from the elixir template
  # Only task-specific config needed now!
  task "lint" do
    from elixir
    run "mix credo --strict"
  end

  task "test" do
    from elixir
    env "MIX_ENV", "test"  # Adds to template env
    run "mix test"
  end

  # Override template settings when needed
  task "build" do
    from elixir
    env "MIX_ENV", "prod"
    run "mix release"
    output "release", "_build/prod/rel/myapp"
    after_ ["lint", "test"]
  end

  # Different template for JS tasks
  task "docs" do
    from node
    run "npm run build:docs"
  end
end

# Without templates (repetitive):
#   task "lint" do
#     container "elixir:1.16"
#     mount src, "/app"
#     workdir "/app"
#     ...
#   end
#   task "test" do
#     container "elixir:1.16"
#     mount src, "/app"
#     workdir "/app"
#     ...
#   end
#
# With templates (DRY):
#   elixir = template("elixir") |> template_container("elixir:1.16") |> ...
#   task "lint" do from elixir; run "mix credo"; end
#   task "test" do from elixir; run "mix test"; end
